File: multimedia/MajCover.js
var coverArt = require('cover-art'),
	wget = require('node-wget'),
	glb = require('globalAutoradio'),
	fs = require('fs'),
	DBWrapper = require('node-dbi').DBWrapper,
	path = require('path'),
	child_process = require('child_process'),
	ini = require('ini');

/* ----------------------------------------------------- */
/* lecture fichier de CONF                               */
/* ----------------------------------------------------- */
var fileIni = fs.readFileSync('./data/config.ini', 'utf-8', function(err) {
	if (err) {
		console.log(glb.Time() + ' [MULTIMEDIA]\t ERR ecriture '+err);
	}
});
 
//Création d'un JSON à partir du fichier ini '
var config = ini.parse(fileIni);

/* ----------------------------------------------------- */
/* fonction                                              */
/* ----------------------------------------------------- */
// mise à jour BDD
function majBDD(id, target) {
	
	var db1 = new DBWrapper( 'pg', {
		host     : config.database.host,
	 	user     : config.database.user,
	  	password : config.database.password,
	  	database : config.database.database
	});
	db1.connect( function(err) {

		if (err) {
			console.log(glb.Time() + ' [MAJCOVER]\t Error connecting database DB1...'+err);
			process.exit(1);
		} else {
			console.log(glb.Time() + ' [MAJCOVER]\t DB1 Database is connected ...');
	
			var sql = "UPDATE albums SET coverPath = '"+target+"' WHERE id = "+id+";";
			db1.fetchAll(sql, function(err, result) {
									
				if(err) {
					console.log(glb.Time() + ' [MAJCOVER]\t ERR mise à jour BDD '+err);    
				} else {
		
					if (config.MajCover.verbose) {
						console.log(glb.Time() + ' [MAJCOVER]\t OK mise a jour BDD '+result);
					}
				}
			});
		}
	});
	db1.close();

}	

/* ----------------------------------------------------- */
/* recuperation liste des album par artistes             */
/* ----------------------------------------------------- */
// recuperation des Album par artists
db = new DBWrapper( 'pg', {
	host     : config.database.host,
 	user     : config.database.user,
  	password : config.database.password,
  	database : config.database.database
});
db.connect( function(err) {

	if (err) {
		console.log(glb.Time() + ' [MAJCOVER]\t Error connecting database ...'+err);
		process.exit(1);
	} else {
		console.log(glb.Time() + ' [MAJCOVER]\t Database is connected ...');
	}

});

db.fetchAll('SELECT albums.id, artists.nomArtist, albums.nomAlbum FROM albums '+
				'INNER JOIN artists on artists.id = albums.artistsID ', function(err, result) {
	
	if(err) {
		console.log(glb.Time() + ' [MAJCOVER]\t ERR recuperation albums '+err);    
	} else {
		
		if (config.MajCover.verbose) {
			console.log(glb.Time() + ' [MAJCOVER]\t recuperation albums '+result);
		}
		
		result.forEach( function(row, index, array) {
		
			Artist = glb.Replace(row['nomartist']);
			Album = glb.Replace(row['nomalbum']);
		
			var imgTarget = config.MajCover.repCover+"/"+Artist+"."+Album+".png"
			
			// test si le cover est déjà présent
			
			fs.exists(imgTarget, function(exist) {

				if (exist) {
					
					if (config.MajCover.verbose) {
						console.log(glb.Time() + ' [MAJCOVER]\t fichier '+imgTarget+' existant');
					}
					
					majBDD(row['id'], imgTarget);
				
				} else {
						
					if (config.MajCover.verbose) {
						console.log(glb.Time() + ' [MAJCOVER]\t fichier '+imgTarget+' absent, telechargement');
					}		
					
					// recherche album et telechargement
					coverArt(row['nomartist'], row['nomalbum'], 'extralarge', function (err, url) {
   	
   						// si on trouve le cover, on le charge
						if ( url ) {
						
							if (config.MajCover.verbose) {
								console.log(glb.Time() + ' [MAJCOVER]\t fichier '+imgTarget+' absent, WGET');
							}
						
							wget(url, function() {
							
								if (config.MajCover.verbose) {
									console.log(glb.Time() + ' [MAJCOVER]\t telechargement termine');
									console.log(glb.Time() + ' [MAJCOVER]\t rename '+path.basename(url)+' '+imgTarget);
								}
								fs.rename(path.basename(url), imgTarget, function(err) {
						
									if (err) {
										console.log(glb.Time() + ' [MULTIMEDIA]\t ERR ecriture '+err);
									}
									
									// creation thumbnail
									var convert = child_process.exec('convert '+imgTarget+' -resize 50x50 '+imgTarget+'.THUMB');
									convert.stdout.on('data', function (data) {
										console.log('INFO CONVERT '+data);	
									});
									convert.stderr.on('data', function (data) {
										console.log('ERROR CONVERT '+data);	
									});
									
								});
								
								majBDD(row['id'], imgTarget);

							});
						
						// si on ne trouve pas le cover, on laisse le defaut
						} else {
						
							if (config.MajCover.verbose) {
								console.log(glb.Time() + ' [MAJCOVER]\t album non trouvé, mise en place defCover');
							}
							
						}
						
					});
				}		
			});

		console.log(index);
		});
		console.log('END');
	}
});


// Photo artist seul
//coverArt('The Beatles', function (err, url) {
//    console.log(url);
//    wget(url);
    //=> http://path/to/beatles.jpg 
//});

File: multimedia/MajCover.py
# recuperation des cover
#      utilisation de "sacad"

import psycopg2
import os
import unicodedata
import subprocess, re
from subprocess import call

# mise en place de l'encoding UTF8
import sys
reload(sys)  
sys.setdefaultencoding('utf8')

# configuration
REPCover = "data/cover"

# initialisation de la base de donnee
print "initialisation database bibliotheque"
try:
    db = psycopg2.connect("dbname='db_media' user='pi' host='127.0.0.1' password='muscadet'")
    cursor = db.cursor()
except:
    print "I am unable to connect to the database"
    exit(1)

print "Database connectee"

# function de mise a jour du cover en base
def majCoverBDD(pathCover,albumID):

	print "--> mise a jour cover"
	try:
		cursor.execute("""UPDATE albums SET coverPath = %s WHERE id = %s;""", (pathCover,albumID,) )
		db.commit()
	except:
		print "--> ERREUR mise a jour cover"
		exit(1)

# on recupere la liste des albums
cursor.execute("""SELECT albums.id as id, artists.nomArtist as artist, albums.nomAlbum as album FROM albums INNER JOIN artists on artists.id = albums.artistsID""")
rows = cursor.fetchall()
for lig in rows:
	
	albumID = lig[0]
	artist = lig[1]
	album = lig[2]
	
	# suppression des accents
	tmpcover = unicode(artist+"."+album+".jpg", 'utf-8')
	cover = unicodedata.normalize('NFD', tmpcover).encode('ascii', 'ignore')
	cover = cover.replace(' ','_')
	cover = cover.replace('/','-')
	cover = cover.replace('\'','-')
	pathCover = REPCover+'/'+cover
	
	print "traitement : ",artist," ",album," ",cover
	
	# on verifie la presence du cover
	if os.path.exists(pathCover):
		print 'cover existant, on met a jour la BDD'
		majCoverBDD(pathCover,albumID)
		
	else:
	
		# recherche du cover
		print 'cover absent, recherche du cover'
		cmd = ['sacad', artist, album, '600', pathCover]
		p = subprocess.call(cmd, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
		
		# maj BDD si cover downloaded
		if os.path.exists(pathCover):
			majCoverBDD(pathCover,albumID)
